### PR TITLE
Made rounding in convert_image_dtype for numbers close to zero

### LINKED ISSUE #48701: Function used in many augmentations (convert_image_dtype) has an issue
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab default
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Colab default
- TensorFlow version (use command below): v2.4.1-0-g85c8b2a817f 2.4.1
- Python version: 3
- Bazel version (if compiling from source): No
- GCC/Compiler version (if compiling from source): No
- CUDA/cuDNN version: No
- GPU model and memory: No

**Describe the current behavior**
There is a function convert_image_dtype that used in many augmentation ops like tf.image.stateless_random_brightness and etc.
That function has a rounding issue here https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/ops/image_ops_impl.py#L2290
Maybe other cases also have this bug (not checked).

When we casting float to int, we should use rounding. In convert_image_dtype rounding implemented as shift by 0.5 which is correct for max value, but is not correct for 0.
See example by link below.

**Describe the expected behavior**
Every time convert_image_dtype changes value scale following casting to int, it should use rounding before casting.

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1xZqyuAlZu_xkDZZHNsr7K8g6MyapcNPi?usp=sharing

### PR DESCRIPTION
**Pull request by shkarupa-alex**
Fixes https://github.com/tensorflow/tensorflow/issues/48701

That issue already was fixed by https://github.com/tensorflow/tensorflow/pull/49868
But after merging all changes were overwritten in https://github.com/tensorflow/tensorflow/commit/56ab2308f72da337865cf765a1844ed9e990d02e#diff-a5a22434f0c18768fc2e10c0e0420ac6f111a5802f67e3df54f155bfefc7094f and now issue raises again

### CODE REVIEW

**Comment by shkarupa-alex** [comment]
Any update on merging this PR?

**Comment by edloper** [comment]
I'm not familiar with this code, so I'm probably not the right person to review it.  

But it looks like the original change was automatically rolled back because it caused a test in this file to fail: https://github.com/tensorflow/tensorflow/blob/v2.8.0/tensorflow/compiler/tests/image_ops_test.py.  (Unfortunately, since it was rolled back several months ago, the logs from the failed test aren't available anymore.)

**Comment by gbaned** [comment]
@rohan100jain Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
@rohan100jain Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @rohan100jain Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @jsmeredith Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @jsmeredith Can you please review this PR ? Thank you!

**Comment by cantonios** [comment]
@shkarupa-alex have you addressed the compiler test failure?

**Comment by shkarupa-alex** [comment]
@cantonios , i don't see any test failures

**Comment by gbaned** [comment]
Hi @cantonios Can you please assist on above comments from @shkarupa-alex. Thank you!

**Review by cantonios — APPROVED** [review]

File: tensorflow/python/ops/image_ops_test.py [hunk 1]
```diff
@@ -965,7 +965,7 @@ def testBatchSaturation(self):
     x_np = np.array(x_data, dtype=np.uint8).reshape(x_shape)
 
     saturation_factor = 0.5
-    y_data = [6, 9, 13, 140, 180, 226, 135, 121, 234, 172, 255, 128]
+    y_data = [7, 9, 13, 140, 181, 226, 135, 121, 234, 172, 255, 128]
```
Comment (cantonios) [inline tensorflow/python/ops/image_ops_test.py:968]: This is also failing, same as above.

**Review by cantonios — CHANGES_REQUESTED** [review]

**Review by cantonios — CHANGES_REQUESTED** [review]
See above mentioned failure

**Comment by gbaned** [comment]
Hi @shkarupa-alex Can you please check @cantonios's comments and keep us posted ? Thank you!

**Comment by gbaned** [comment]
Hi @shkarupa-alex Any update on this PR? Please. Thank you!

**Comment by shkarupa-alex** [comment]
I'm not sure how to fix that, due to different behavior with CPU and GPU.
Current tests work well with CPU. But with GPU fail with error (difference in 1 unit for some values).

**Comment by shkarupa-alex** [comment]
The issue comes from rescaling between dtypes
When using multiplication with inverse scale GPU and CPU ops produce different results.
Here is a short example:

```python
import numpy as np
import tensorflow as tf
from tensorflow.python.ops import gen_image_ops


def convert_image_dtype(image, dtype, saturate=False, invscale=True, name=None):
    cast = tf.cast(image, dtype)
    
    if invscale:
        return tf.math.multiply(cast, 1. / image.dtype.max, name=name)
    else:
        return tf.math.divide(tf.cast(image, dtype), 255.)


def adjust_saturation(image, saturation_factor, invscale=True, name=None):
    flt_image = convert_image_dtype(image, tf.dtypes.float32, invscale=invscale)
    adjusted = gen_image_ops.adjust_saturation(flt_image, saturation_factor)

    return adjusted
    
def test(invscale=True):
    x_shape = [2, 2, 3]
    x_data = [0, 5, 13, 54, 135, 226, 37, 8, 234, 90, 255, 1]
    x_np = np.array(x_data, dtype=np.uint8).reshape(x_shape)

    saturation_factor = 0.5
    y_data = [7, 9, 13, 140, 181, 226, 135, 121, 234, 172, 255, 128]
    y_np = np.array(y_data, dtype=np.uint8).reshape(x_shape)

    x = tf.constant(x_np, shape=x_shape)
    # x = tf.cast(x, 'float32') / 255.
    y = adjust_saturation(x, saturation_factor, invscale=invscale)
    y_tf = y.numpy()
    print(y_tf[0, 1, 1] * 255.)


print('With inv scale')
with tf.device("/cpu:0"):
    test()
with tf.device("/gpu:0"):
    test()
    
print('\nWithout inv scale')
with tf.device("/cpu:0"):
    test(invscale=False)
with tf.device("/gpu:0"):
    test(invscale=False)
```
and it's output:
```
With inv scale
180.50001204013824
180.49999684095383

Without inv scale
180.49999684095383
180.49999684095383
```

**Comment by cantonios** [comment]
You should change the test so it is not so sensitive to these numerical perturbations - essentially, avoid results that are expected to be at `X.5` when math+rounding is involved.

**Comment by gbaned** [comment]
Hi @shkarupa-alex Can you please check @cantonios's comments and keep us posted ? Thank you!

**Comment by gbaned** [comment]
Hi @shkarupa-alex Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @shkarupa-alex Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @shkarupa-alex  I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen! Thank you for your contribution!

**Pull request closed without merging**